dartagnan/src/main/java/com/dat3m/dartagnan/expression/integers/IntExtract.java

```diff
@@ -0,0 +1,35 @@
+package com.dat3m.dartagnan.expression.integers;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionKind;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.base.UnaryExpressionBase;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
+import com.dat3m.dartagnan.expression.utils.ExpressionHelper;
+import com.google.common.base.Preconditions;
+
+public class IntExtract extends UnaryExpressionBase<IntegerType, ExpressionKind.Other> {
+
+    private final int lowBit;
+    private final int highBit;
+
+    public IntExtract(Expression operand, int lowBit, int highBit) {
+        super(TypeFactory.getInstance().getIntegerType(highBit - lowBit + 1),
+                ExpressionKind.Other.BV_EXTRACT,
+                operand);
+        ExpressionHelper.checkExpectedType(operand, IntegerType.class);
+        int originalWidth = ((IntegerType)operand.getType()).getBitWidth();
+        Preconditions.checkArgument(0 <= lowBit && lowBit <= highBit && highBit < originalWidth);
```

Comment:
Can extract be used as the identity? i.e. `highBit <= originalWidth`

Reply:
It can be the identity, but that is when "highBit = originalWidth - 1" which is valid.
Notice that the bits indicate 0-based positions whereas sizes start from 1, so there is this off-by-one mismatch.